src/linked_hash_map.rs

```diff
@@ -0,0 +1,540 @@
+// Copyright 2013 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+
+//! A HashMap wrapper that holds key-value pairs in insertion order.
+//!
+//! # Example
+//!
+//! ```rust
+//! use collect::LinkedHashMap;
+//!
+//! let mut map: LinkedHashMap<int, int> = LinkedHashMap::new();
+//! map.insert(1, 10);
+//! map.insert(2, 20);
+//! map.insert(3, 30);
+//! assert_eq!(*map.get(&1).unwrap(), 10);
+//! assert_eq!(*map.get(&2).unwrap(), 20);
+//! assert_eq!(*map.get(&3).unwrap(), 30);
+//! ```
+
+use std::cmp::{PartialEq, Eq};
+use std::collections::HashMap;
+use std::fmt;
+use std::hash::Hash;
+use std::iter::{Iterator, Extend};
+use std::mem;
+use std::ptr;
+
+// FIXME(conventions): implement indexing?
+
+struct KeyRef<K> { k: *const K }
+
+struct LinkedHashMapEntry<K, V> {
+    next: *mut LinkedHashMapEntry<K, V>,
+    prev: *mut LinkedHashMapEntry<K, V>,
+    key: K,
+    value: V,
+}
+
+/// A linked hash map.
+pub struct LinkedHashMap<K, V> {
+    map: HashMap<KeyRef<K>, Box<LinkedHashMapEntry<K, V>>>,
+    head: *mut LinkedHashMapEntry<K, V>,
```

Comment:
Not sure about Deque. However, I'd like to implement `into_iter` one day (it is a somewhat harder than plain `iter`).